File: Monotonic Stack/Q1. Final Prices With a Special Discount in a Shop/Q1_FinalPricesWithaSpecialDiscountinaShop.java
// Solution class
class Solution {

    // Method to calculate final prices after discount
    public int[] finalPrices(int[] prices) {

        // Length of the array
        int n = prices.length;

        // Result array to store final prices
        int[] res = new int[n];

        // Loop through each item
        for (int i = 0; i < n; i++) {

            // Initialize result with original price
            res[i] = prices[i];

            // Check next items for discount
            for (int j = i + 1; j < n; j++) {

                // If next price is less than or equal to current
                if (prices[j] <= prices[i]) {

                    // Apply discount
                    res[i] -= prices[j];

                    // Stop after first eligible discount
                    break;
                }
            }
        }

        // Return final prices
        return res;
    }
}

File: Linear Shoal/Monotonic Stack/Q1. Final Prices With a Special Discount in a Shop/Q1_FinalPricesWithaSpecialDiscountinaShop.py
# Solution class
class Solution:

    # Function to calculate final prices after discount
    def finalPrices(self, prices):

        # Length of the array
        n = len(prices)

        # nge[i] stores index of next smaller or equal price for prices[i]
        nge = [-1] * n

        # Stack to keep track of indices while scanning
        st = []

        # Start from the last index
        st.append(n - 1)

        # Traverse array from second last to first
        for i in range(n - 2, -1, -1):

            # Pop indices whose prices are greater than current price
            while st and prices[st[-1]] > prices[i]:
                st.pop()

            # If stack not empty, next smaller/equal price exists
            if st:
                nge[i] = st[-1]

            # Push current index onto stack
            st.append(i)

        # Array to store final discounted prices
        ans = [0] * n

        # Calculate final price for each item
        for i in range(n):

            # If no smaller/equal price ahead, price stays same
            if nge[i] == -1:
                ans[i] = prices[i]

            # Else subtract the discount price
            else:
                ans[i] = prices[i] - prices[nge[i]]

        # Return the final prices
        return ans
